File: go/identity.go
package spiffile

import (
	"crypto/ecdsa"
	"crypto/elliptic"
	"encoding/base64"
	"encoding/json"
	"fmt"
	"math/big"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"

	"github.com/golang-jwt/jwt/v5"
)

// Defaults shared by all spiffile implementations.
const (
	DefaultTTL    = 60 * time.Second
	DefaultLeeway = 30 * time.Second
)

// Environment configuration, shared by all spiffile implementations.
const (
	EnvDir        = "SPIFFILE_DIR"
	EnvIDFile     = "SPIFFILE_ID_FILE"
	EnvKeyFile    = "SPIFFILE_KEY_FILE"
	EnvBundleFile = "SPIFFILE_BUNDLE_FILE"

	DirIDFilename     = "id"
	DirKeyFilename    = "key.pem"
	DirBundleFilename = "bundle.json"
)

// Caller is the verified peer of an inbound request.
type Caller struct {
	ID     SpiffeID
	Claims jwt.MapClaims
}

// fileKeySource reads the private key from a file, hot-reloading when it
// changes. Key rotation replaces the key file; signing must pick the new key
// up without a process restart — otherwise tokens break once the rotated-out
// public key is pruned from peers' bundles.
type fileKeySource struct {
	path string

	mu    sync.Mutex
	mtime int64
	size  int64
	key   *ecdsa.PrivateKey
	kid   string
}

func (s *fileKeySource) get() (*ecdsa.PrivateKey, string, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	info, err := os.Stat(s.path)
	if err != nil {
		return nil, "", fmt.Errorf("cannot stat key at %s: %w", s.path, err)
	}
	if s.key == nil || info.ModTime().UnixNano() != s.mtime || info.Size() != s.size {
		pemBytes, err := os.ReadFile(s.path)
		if err != nil {
			return nil, "", fmt.Errorf("cannot read key at %s: %w", s.path, err)
		}
		key, jwk, err := parsePrivateKeyPEM(pemBytes)
		if err != nil {
			return nil, "", fmt.Errorf("cannot parse key at %s: %w", s.path, err)
		}
		s.key = key
		s.kid = jwk.Kid
		s.mtime = info.ModTime().UnixNano()
		s.size = info.Size()
	}
	return s.key, s.kid, nil
}

// Identity is a service's own identity plus the trust bundle to verify
// peers against.
type Identity struct {
	ID SpiffeID

	keySource    *fileKeySource
	bundleSource *FileBundleSource
}

// FromFiles loads an identity from the three profile files.
func FromFiles(idFile, keyFile, bundleFile string) (*Identity, error) {
	raw, err := os.ReadFile(idFile)
	if err != nil {
		return nil, fmt.Errorf("cannot read identity file: %w", err)
	}
	id, err := ParseSpiffeID(strings.TrimSpace(string(raw)))
	if err != nil {
		return nil, err
	}
	identity := &Identity{
		ID:           id,
		keySource:    &fileKeySource{path: keyFile},
		bundleSource: NewFileBundleSource(bundleFile),
	}
	if _, _, err := identity.keySource.get(); err != nil {
		return nil, err
	}
	return identity, nil
}

// FromEnv loads an identity from environment configuration: either
// SPIFFILE_DIR (containing id / key.pem / bundle.json) or the three explicit
// file variables, which take precedence per variable.
func FromEnv() (*Identity, error) {
	dir := os.Getenv(EnvDir)
	idFile := os.Getenv(EnvIDFile)
	keyFile := os.Getenv(EnvKeyFile)
	bundleFile := os.Getenv(EnvBundleFile)
	if dir != "" {
		if idFile == "" {
			idFile = filepath.Join(dir, DirIDFilename)
		}
		if keyFile == "" {
			keyFile = filepath.Join(dir, DirKeyFilename)
		}
		if bundleFile == "" {
			bundleFile = filepath.Join(dir, DirBundleFilename)
		}
	}
	if idFile == "" || keyFile == "" || bundleFile == "" {
		return nil, fmt.Errorf("identity not configured: set %s, or %s + %s + %s",
			EnvDir, EnvIDFile, EnvKeyFile, EnvBundleFile)
	}
	return FromFiles(idFile, keyFile, bundleFile)
}

// Token mints a JWT-SVID for a single target audience.
func (i *Identity) Token(audience string, ttl time.Duration) (string, error) {
	audienceID, err := ParseSpiffeID(audience)
	if err != nil {
		return "", fmt.Errorf("invalid audience: %w", err)
	}
	if ttl <= 0 {
		ttl = DefaultTTL
	}
	key, kid, err := i.keySource.get()
	if err != nil {
		return "", err
	}
	now := time.Now()
	token := jwt.NewWithClaims(jwt.SigningMethodES256, jwt.MapClaims{
		"sub": i.ID.String(),
		"aud": audienceID.String(),
		"iat": now.Unix(),
		"exp": now.Add(ttl).Unix(),
	})
	token.Header["kid"] = kid
	return token.SignedString(key)
}

// Verify verifies an inbound JWT-SVID and returns the verified caller.
// An empty audience defaults to this identity's own SPIFFE ID.
func (i *Identity) Verify(tokenString, audience string, leeway time.Duration) (Caller, error) {
	expectedAudience := audience
	if expectedAudience == "" {
		expectedAudience = i.ID.String()
	}
	if leeway <= 0 {
		leeway = DefaultLeeway
	}

	// Read the claimed identity (unverified) to scope the key lookup —
	// the security pivot of the profile: only keys bound to the claimed
	// identity in the bundle may validate it.
	unverified, _, err := jwt.NewParser().ParseUnverified(tokenString, jwt.MapClaims{})
	if err != nil {
		return Caller{}, fmt.Errorf("malformed token: %w", err)
	}
	claimedSub, err := unverified.Claims.GetSubject()
	if err != nil || claimedSub == "" {
		return Caller{}, fmt.Errorf("token has no sub claim")
	}
	claimedID, err := ParseSpiffeID(claimedSub)
	if err != nil {
		return Caller{}, fmt.Errorf("token sub is not a SPIFFE ID: %w", err)
	}

	// The profile requires aud to be a single string. golang-jwt would
	// happily match the expected audience inside an array, so reject
	// non-string aud claims here.
	if aud, present := unverified.Claims.(jwt.MapClaims)["aud"]; present {
		if _, isString := aud.(string); !isString {
			return Caller{}, fmt.Errorf("token rejected: aud must be a single string audience")
		}
	}

	bundle, err := i.bundleSource.Get()
	if err != nil {
		return Caller{}, err
	}
	boundJWKs, err := bundle.KeysFor(claimedID.String())
	if err != nil {
		return Caller{}, err
	}

	kid, _ := unverified.Header["kid"].(string)
	candidates := candidateKeys(boundJWKs, kid)
	if len(candidates) == 0 {
		return Caller{}, fmt.Errorf("no usable keys bound to %q", claimedSub)
	}

	var lastErr error
	for _, publicKey := range candidates {
		claims := jwt.MapClaims{}
		_, err := jwt.NewParser(
			jwt.WithValidMethods([]string{Algorithm}),
			jwt.WithAudience(expectedAudience),
			jwt.WithLeeway(leeway),
			jwt.WithExpirationRequired(),
		).ParseWithClaims(tokenString, claims, func(*jwt.Token) (any, error) {
			return publicKey, nil
		})
		if err == nil {
			return Caller{ID: claimedID, Claims: claims}, nil
		}
		lastErr = err
		if !strings.Contains(err.Error(), "signature is invalid") {
			// Signature was fine but a claim failed — no other key will fix that.
			return Caller{}, fmt.Errorf("token rejected: %w", err)
		}
	}
	return Caller{}, fmt.Errorf("signature does not match any key bound to %q: %w", claimedSub, lastErr)
}

// candidateKeys parses the JWKs bound to an identity into public keys,
// preferring an exact kid match and falling back to all bound keys.
func candidateKeys(jwks []json.RawMessage, kid string) []*ecdsa.PublicKey {
	var matched, all []*ecdsa.PublicKey
	for _, raw := range jwks {
		var jwk JWK
		if err := json.Unmarshal(raw, &jwk); err != nil {
			continue
		}
		publicKey, err := jwkToPublicKey(jwk)
		if err != nil {
			continue
		}
		all = append(all, publicKey)
		if kid != "" && jwk.Kid == kid {
			matched = append(matched, publicKey)
		}
	}
	if len(matched) > 0 {
		return matched
	}
	return all
}

func jwkToPublicKey(jwk JWK) (*ecdsa.PublicKey, error) {
	if jwk.Kty != "EC" || jwk.Crv != "P-256" {
		return nil, fmt.Errorf("unsupported key type %s/%s", jwk.Kty, jwk.Crv)
	}
	x, err := base64.RawURLEncoding.DecodeString(jwk.X)
	if err != nil {
		return nil, err
	}
	y, err := base64.RawURLEncoding.DecodeString(jwk.Y)
	if err != nil {
		return nil, err
	}
	return &ecdsa.PublicKey{
		Curve: elliptic.P256(),
		X:     new(big.Int).SetBytes(x),
		Y:     new(big.Int).SetBytes(y),
	}, nil
}

File: ts/src/identity.ts
/**
 * Workload identity: mint and verify JWT-SVIDs from file-delivered material.
 * Zero dependencies — JWS compact serialization via node:crypto (ES256 with
 * IEEE P1363 signatures, as JOSE requires).
 */
import { createPublicKey, sign as cryptoSign, verify as cryptoVerify, type KeyObject } from "node:crypto"
import { readFileSync, statSync } from "node:fs"
import { join } from "node:path"

import { Bundle, FileBundleSource } from "./bundle.js"
import { InvalidTokenError, SpiffileError } from "./errors.js"
import { ALGORITHM, privateKeyFromPem, publicJwk, type Jwk } from "./keys.js"
import { SpiffeId } from "./spiffeId.js"

export const DEFAULT_TTL_SECONDS = 60
export const DEFAULT_LEEWAY_SECONDS = 30

/** Profile v0: ES256 only. A verifier MUST reject anything else. */
export const ALLOWED_ALGORITHMS = [ALGORITHM]

export const ENV_DIR = "SPIFFILE_DIR"
export const ENV_ID_FILE = "SPIFFILE_ID_FILE"
export const ENV_KEY_FILE = "SPIFFILE_KEY_FILE"
export const ENV_BUNDLE_FILE = "SPIFFILE_BUNDLE_FILE"

export const DIR_ID_FILENAME = "id"
export const DIR_KEY_FILENAME = "key.pem"
export const DIR_BUNDLE_FILENAME = "bundle.json"

/** The verified peer of an inbound request. */
export interface Caller {
  id: SpiffeId
  claims: Record<string, unknown>
}

const b64url = (data: Buffer | string): string => Buffer.from(data).toString("base64url")

/**
 * Reads the private key from a file, hot-reloading when it changes.
 * Key rotation replaces the key file; signing must pick the new key up
 * without a process restart — otherwise tokens break once the rotated-out
 * public key is pruned from peers' bundles.
 */
class FileKeySource {
  private stamp: string | null = null
  private key: KeyObject | null = null
  private kid = ""

  constructor(private readonly path: string) {}

  get(): { key: KeyObject; kid: string } {
    const stat = statSync(this.path)
    const stamp = `${stat.mtimeMs}:${stat.size}`
    if (this.key === null || stamp !== this.stamp) {
      this.key = privateKeyFromPem(readFileSync(this.path))
      this.kid = publicJwk(this.key).kid
      this.stamp = stamp
    }
    return { key: this.key, kid: this.kid }
  }
}

class StaticKeySource {
  private readonly key: KeyObject
  private readonly kid: string

  constructor(key: KeyObject) {
    this.key = key
    this.kid = publicJwk(key).kid
  }

  get(): { key: KeyObject; kid: string } {
    return { key: this.key, kid: this.kid }
  }
}

/** A service's own identity plus the trust bundle to verify peers against. */
export class Identity {
  readonly id: SpiffeId
  private readonly keySource: FileKeySource | StaticKeySource
  private readonly bundleSource: FileBundleSource

  constructor(id: SpiffeId, privateKey: KeyObject | FileKeySource, bundleSource: FileBundleSource) {
    this.id = id
    this.keySource = privateKey instanceof FileKeySource ? privateKey : new StaticKeySource(privateKey)
    this.bundleSource = bundleSource
  }

  // -- construction --------------------------------------------------------

  static fromFiles(idFile: string, keyFile: string, bundleFile: string): Identity {
    const spiffeId = SpiffeId.parse(readFileSync(idFile, "utf8").trim())
    return new Identity(spiffeId, new FileKeySource(keyFile), new FileBundleSource(bundleFile))
  }

  /**
   * Load identity from environment configuration: either SPIFFILE_DIR (a
   * directory containing id, key.pem and bundle.json) or the three explicit
   * file variables, which take precedence per variable.
   */
  static fromEnv(env: Record<string, string | undefined> = process.env): Identity {
    const dir = env[ENV_DIR]
    const idFile = env[ENV_ID_FILE] ?? (dir ? join(dir, DIR_ID_FILENAME) : undefined)
    const keyFile = env[ENV_KEY_FILE] ?? (dir ? join(dir, DIR_KEY_FILENAME) : undefined)
    const bundleFile = env[ENV_BUNDLE_FILE] ?? (dir ? join(dir, DIR_BUNDLE_FILENAME) : undefined)
    if (!idFile || !keyFile || !bundleFile) {
      throw new SpiffileError(
        `identity not configured: set ${ENV_DIR}, or ${ENV_ID_FILE} + ${ENV_KEY_FILE} + ${ENV_BUNDLE_FILE}`,
      )
    }
    return Identity.fromFiles(idFile, keyFile, bundleFile)
  }

  // -- outbound -------------------------------------------------------------

  /** Mint a JWT-SVID for a single target audience. */
  token(audience: string | SpiffeId, ttl: number = DEFAULT_TTL_SECONDS): string {
    const audienceId = audience instanceof SpiffeId ? audience : SpiffeId.parse(audience)
    const { key, kid } = this.keySource.get()
    const now = Math.floor(Date.now() / 1000)
    const header = { alg: ALGORITHM, typ: "JWT", kid }
    const claims = {
      sub: this.id.toString(),
      aud: audienceId.toString(),
      iat: now,
      exp: now + ttl,
    }
    const signingInput = `${b64url(JSON.stringify(header))}.${b64url(JSON.stringify(claims))}`
    const signature = cryptoSign("sha256", Buffer.from(signingInput), { key, dsaEncoding: "ieee-p1363" })
    return `${signingInput}.${signature.toString("base64url")}`
  }

  // -- inbound --------------------------------------------------------------

  /**
   * Verify an inbound JWT-SVID and return the verified caller.
   * `audience` defaults to this identity's own SPIFFE ID.
   *
   * Throws UnknownIdentityError if the claimed identity has no keys in the
   * bundle, InvalidTokenError for everything else.
   */
  verify(token: string, audience?: string | SpiffeId, leeway: number = DEFAULT_LEEWAY_SECONDS): Caller {
    const expectedAudience = (audience ?? this.id).toString()

    const parts = token.split(".")
    if (parts.length !== 3) {
      throw new InvalidTokenError("malformed token: expected three dot-separated segments")
    }
    let header: { alg?: string; kid?: string }
    let claims: Record<string, unknown>
    try {
      header = JSON.parse(Buffer.from(parts[0], "base64url").toString("utf8"))
      claims = JSON.parse(Buffer.from(parts[1], "base64url").toString("utf8"))
    } catch (error) {
      throw new InvalidTokenError(`malformed token: ${error}`)
    }

    // Algorithm pinning: the allow-list decides, never the token header.
    if (!header.alg || !ALLOWED_ALGORITHMS.includes(header.alg)) {
      throw new InvalidTokenError(`token rejected: algorithm ${header.alg} is not allowed`)
    }
    const claimedSub = claims.sub
    if (typeof claimedSub !== "string" || !claimedSub) {
      throw new InvalidTokenError("token has no sub claim")
    }
    const claimedId = SpiffeId.parse(claimedSub)

    // The profile requires aud to be a single string; strict equality also
    // rejects array audiences, even ones containing the expected value.
    if (claims.aud !== expectedAudience) {
      throw new InvalidTokenError("token rejected: audience doesn't match")
    }
    const now = Date.now() / 1000
    if (typeof claims.exp !== "number") {
      throw new InvalidTokenError("token rejected: missing exp claim")
    }
    if (claims.exp + leeway < now) {
      throw new InvalidTokenError("token rejected: expired")
    }
    if (typeof claims.iat === "number" && claims.iat - leeway > now) {
      throw new InvalidTokenError("token rejected: issued in the future")
    }

    // The security pivot of the profile: only keys bound to the claimed
    // identity in the bundle may validate it. May throw UnknownIdentityError.
    const boundJwks = this.bundleSource.get().keysFor(claimedId.toString())
    const candidates = boundJwks.filter((jwk) => jwk.kid === header.kid)
    const signature = Buffer.from(parts[2], "base64url")
    const signingInput = Buffer.from(`${parts[0]}.${parts[1]}`)

    for (const jwk of candidates.length > 0 ? candidates : boundJwks) {
      let publicKey: KeyObject
      try {
        publicKey = createPublicKey({ key: { kty: jwk.kty, crv: jwk.crv, x: jwk.x, y: jwk.y }, format: "jwk" })
      } catch {
        continue
      }
      if (cryptoVerify("sha256", signingInput, { key: publicKey, dsaEncoding: "ieee-p1363" }, signature)) {
        return { id: claimedId, claims }
      }
    }
    throw new InvalidTokenError(`signature does not match any key bound to '${claimedSub}'`)
  }
}

/**
 * Read a JWT-SVID's `aud` claim WITHOUT verifying the signature.
 *
 * For routing and diagnostics only — e.g. choosing which trust context to
 * verify under when the caller has not stated an expected audience. The result
 * is attacker-controlled; never use it for an access decision, always follow
 * up with {@link Identity.verify}.
 *
 * Returns `null` when the token carries no `aud`. Rejects a non-string (e.g.
 * array) `aud` so callers don't silently treat a multi-audience token as
 * single-audience.
 */
export function unverifiedAudience(token: string): string | null {
  const parts = token.split(".")
  if (parts.length !== 3) {
    throw new InvalidTokenError("malformed token: expected three dot-separated segments")
  }
  let claims: Record<string, unknown>
  try {
    claims = JSON.parse(Buffer.from(parts[1], "base64url").toString("utf8"))
  } catch (error) {
    throw new InvalidTokenError(`malformed token: ${error}`)
  }
  const aud = claims.aud
  if (aud === undefined || aud === null) {
    return null
  }
  if (typeof aud !== "string") {
    throw new InvalidTokenError("aud must be a single string audience")
  }
  return aud
}

export { FileKeySource }

/** Re-exported for provisioning and tests. */
export type { Bundle }
